src/info.rs

```diff
@@ -0,0 +1,692 @@
+use wasm_encoder::SectionId;
+
+pub mod types_interner;
+
+use std::collections::BTreeMap;
+use std::convert::TryFrom;
+use types_interner::{EntityType, InstanceType, Type, TypeId, TypesInterner};
+
+/// A collection of info about modules within a module linking bundle.
+pub(crate) struct ModuleContext<'a> {
+    arena: Vec<ModuleInfo<'a>>,
+    types: TypesInterner<'a>,
+}
+
+impl<'a> ModuleContext<'a> {
+    /// Construct a new `ModuleContext`, pre-populated with an empty root
+    /// module.
+    pub fn new() -> Self {
+        Self {
+            arena: vec![ModuleInfo::Defined(DefinedModuleInfo::default())],
+            types: TypesInterner::default(),
+        }
+    }
+
+    /// Get the root module.
+    pub fn root(&self) -> Module {
+        Module { id: 0 }
+    }
+
+    /// Get the interned types set for this module context.
+    pub fn types(&self) -> &TypesInterner<'a> {
+        &self.types
+    }
+
+    /// Does this context represent a single Wasm module that doesn't use module
+    /// linking, or does it represent a bundle of one or more Wasm modules that
+    /// use module linking?
+    pub fn uses_module_linking(&self) -> bool {
+        self.arena.len() > 1
+            || self.root().initial_sections(self).any(|s| {
+                s.id == SectionId::Alias.into()
+                    || s.id == SectionId::Module.into()
+                    || s.id == SectionId::Instance.into()
+            })
+    }
+
+    /// Get a shared reference to the `DefinedModuleInfo` for this module,
+    /// following through aliases.
+    fn resolve(&self, module: Module) -> &DefinedModuleInfo<'a> {
```

Comment:
Perhaps `defined` or change the below method to `resolve_mut`?

(to name the same)